src/target/at32f43x.c

```diff
@@ -0,0 +1,373 @@
+	uint32_t status = FLASH_STS_OBF;
+	/* Checking for ODF/EOP requires methodically clearing the ODF */
+	while (!(status & FLASH_STS_ODF) && (status & FLASH_STS_OBF)) {
+		status = target_mem_read32(target, AT32F435_FLASH_STS + bank_reg_offset);
+		if (target_check_error(target)) {
+			DEBUG_ERROR("Lost communications with target\n");
+			return false;
+		}
+		if (timeout)
+			target_print_progress(timeout);
+	}
+	if (status & FLASH_STS_PRGMERR) {
+		DEBUG_ERROR("at32f43 flash error, STS: 0x%" PRIx32 "\n", status);
+		return false;
+	}
+	return !(status & FLASH_STS_PRGMERR);
+}
+
+static bool at32f43_flash_prepare(target_flash_s *target_flash)
+{
+	target_s *target = target_flash->t;
+	const at32f43_flash_s *const flash = (at32f43_flash_s *)target_flash;
+	const uint32_t bank_reg_offset = flash->bank_reg_offset;
+	return at32f43_flash_unlock(target, bank_reg_offset);
+}
+
+static bool at32f43_flash_done(target_flash_s *target_flash)
+{
+	target_s *target = target_flash->t;
+	const at32f43_flash_s *const flash = (at32f43_flash_s *)target_flash;
+	const uint32_t bank_reg_offset = flash->bank_reg_offset;
+	return at32f43_flash_lock(target, bank_reg_offset);
+}
+
+static bool at32f43_flash_erase(target_flash_s *target_flash, target_addr_t addr, size_t len)
+{
+	target_s *target = target_flash->t;
+	const at32f43_flash_s *const flash = (at32f43_flash_s *)target_flash;
+	const uint32_t bank_reg_offset = flash->bank_reg_offset;
+
+	if (len != target_flash->blocksize) {
+		DEBUG_ERROR(
+			"%s: Requested erase length %zu does not match blocksize %zu!\n", __func__, len, target_flash->blocksize);
+		return false;
+	}
+
+	at32f43_flash_clear_eop(target, bank_reg_offset);
+
+	/* Prepare for page/sector erase */
+	target_mem_write32(target, AT32F435_FLASH_CTRL + bank_reg_offset, FLASH_CTRL_SECERS);
+	/* Select erased sector by its address */
+	target_mem_write32(target, AT32F435_FLASH_ADDR + bank_reg_offset, addr);
+	/* Start sector erase operation */
+	target_mem_write32(target, AT32F435_FLASH_CTRL + bank_reg_offset, FLASH_CTRL_SECERS | FLASH_CTRL_ERSTR);
+
+	/* Datasheet: page erase takes 50ms (typ), 500ms (max) */
+	return at32f43_flash_busy_wait(target, bank_reg_offset, NULL);
+}
+
+static bool at32f43_flash_write(target_flash_s *target_flash, target_addr_t dest, const void *src, size_t len)
+{
+	target_s *target = target_flash->t;
+	const at32f43_flash_s *const flash = (at32f43_flash_s *)target_flash;
+	const uint32_t bank_reg_offset = flash->bank_reg_offset;
+	const align_e psize = ALIGN_WORD;
+
+	/* Write to bank corresponding to flash region */
+
+	at32f43_flash_clear_eop(target, bank_reg_offset);
+
+	target_mem_write32(target, AT32F435_FLASH_CTRL + bank_reg_offset, FLASH_CTRL_FPRGM);
+	cortexm_mem_write_sized(target, dest, src, len, psize);
+
+	/* Datasheet: flash programming takes 50us (typ), 200us (max) */
+	if (!at32f43_flash_busy_wait(target, bank_reg_offset, NULL))
+		return false;
+
+	return true;
```

Comment:
This can still be simplified to:
```suggestion
	return at32f43_flash_busy_wait(target, bank_reg_offset, NULL);
```

There should be a clang-tidy lint for this (though getting it to fire in all similar situations can be a challenge).

Reply:
Simplified as suggested. There was more complicated logic dropped around it.